api/tokens.go

```diff
@@ -148,3 +151,72 @@ func (api *API) RefreshHandler(ctx context.Context, w http.ResponseWriter, req *
+		ResultsChannel: make(chan string, len(usersList.Users)),
+		BackoffSchedule: []time.Duration{
+			1 * time.Second,
+			3 * time.Second,
+			10 * time.Second,
+		},
+		RetryAllowed: true,
+	}
+	// run api.SignOutUsersWorker concurrently
+	go api.SignOutUsersWorker(ctx, globalSignOut, &usersList.Users)
+	return models.NewSuccessResponse(nil, http.StatusAccepted, nil), nil
+}
+
+// SignOutUsersWorker - signs out users globally by invalidating user's refresh token
+func (api *API) SignOutUsersWorker(ctx context.Context, g *models.GlobalSignOut, usersList *[]models.UserParams) {
+	userSignOutRequestData := g.BuildSignOutUserRequest(usersList, &api.UserPoolId)
+	for i := range userSignOutRequestData {
+		for _, backoff := range g.BackoffSchedule {
+			_, err := api.generateGlobalSignOutRequest(userSignOutRequestData[i])
+			if err != nil {
+				responseErr := models.NewCognitoError(ctx, err, "Cognito AdminUserGlobalSignOut request for sign out")
+				if responseErr.Code != models.TooManyRequestsError { // 1. Process all errors other than TooManyRequestsException (429)
+					if g.RetryAllowed { // 2. Attempt one more request to AdminUserGlobalSignOut if GlobalSignOut.RetryAllowed true, else break to next user
+						g.RetryAllowed = false // 3. Set GlobalSignOut.RetryAllowed to false and request AdminUserGlobalSignOut
+						log.Event(ctx, "Error Cognito AdminUserGlobalSignOut:", log.ERROR, log.Error(err))
+						_, retryErr := api.generateGlobalSignOutRequest(userSignOutRequestData[i])
+						if retryErr != nil { // 4. If error response from request received, process it
+							retryResponseErr := models.NewCognitoError(ctx, err, "Cognito AdminUserGlobalSignOut request for sign out")
+							if retryResponseErr.Code != models.TooManyRequestsError { // 4.1 If error is not a TooManyRequestsException (429), reset GlobalSignOut.RetryAllowed and break to next user
+								g.RetryAllowed = true
+								break
+							}
+						} else { // 5. If request successful, reset GlobalSignOut.RetryAllowed and set original response error to nil - username will be added to GlobalSignOut.ResultsChannel
+							g.RetryAllowed = true
+							err = nil
+						}
+					} else { // 6. If GlobalSignOut.RetryAllowed is already false, know that a second request already made, so reset GlobalSignOut.RetryAllowed and break to next user 
+						g.RetryAllowed = true
+						break
+					}
+				}
+			}
+			if err == nil {
+				g.ResultsChannel <- *userSignOutRequestData[i].Username
+				break
+			}
+			log.Event(ctx, "Error Cognito AdminUserGlobalSignOut:", log.ERROR, log.Error(err))
+			time.Sleep(backoff)
+		}
+	}
+	close(g.ResultsChannel)
+}
+
+// generateGlobalSignOutRequest - local routine to generete the global signout request per user
+func (api *API) generateGlobalSignOutRequest(user *cognitoidentityprovider.AdminUserGlobalSignOutInput) (*cognitoidentityprovider.AdminUserGlobalSignOutOutput, error) {
+	response, error := api.CognitoClient.AdminUserGlobalSignOut(user)
```

Comment:
Looks like you could just `api.CognitoClient.AdminUserGlobalSignOut(user)` here

Reply:
Yes - thought it would be cleaner to have an private package function call the api endpoint  and call that, rather than have two separate api calls in the same function.

Reply:
Yeah I wasn't suggesting any change in structure just that those 2 lines could be written as one - Ah I've just noticed my comment was missing the `return`. What I meant o say was:

Looks like you could just ` return api.CognitoClient.AdminUserGlobalSignOut(user) here`

Reply:
👍